### PR TITLE
Resolve inc metric duplicate registration

### PR DESCRIPTION
**Pull request by zhentaocc**
resolves https://github.com/intel-analytics/BigDL/issues/4002
In previous implementation, each call of quantization, a metric will be registered.  
This causes error when we call quantization with a same metric in one script twice.

### previous implementation :
Each metric is registered at runtime, so there can be name conflict.
```python
  self.metric = Metric(
      MyMetric,
      name="%s_%s" % (framework_metric.__name__, type(metric).__name__)
  )
```
### After this fix:
Customized metrics firstly are registered once for all statically.
```python
METRICS('pytorch').register('PytorchINCMetric', PytorchINCMetric)
METRICS('tensorflow').register('TensorflowINCMetric', TensorflowINCMetric)
```
Then at runtime, we modify the real metric inside `INCMetric` by:
```python
PytorchINCMetric.set_metric(metric)
```
And fetch the registered metric by class name:
```python
self.cfg.evaluation.accuracy.metric = {PytorchINCMetric.__name__: None}
```

### Modifications

- [x]  Add default criterion to INC, allowing 0.99 acc drop by default
- [x]  Fix metric registered in second time
- [x]  Add test with metric assigned second time

### CODE REVIEW

**Comment by zhentaocc** [comment]
@TheaperDeng Please check if this resolves your issue. Or any other suggestions on the fix.

**Comment by zhentaocc** [comment]
@TheaperDeng I have a new correct version. Take a look.

**Review by TheaperDeng — APPROVED** [review]
LGTM

**Comment by zhentaocc** [comment]
Jenkins passed.  
http://10.112.231.51:18888/job/ZOO-PR-NanoTests/253/

File: python/nano/src/bigdl/nano/quantization/neural_compressor/metric/__init__.py [hunk 1]
```diff
@@ -13,10 +13,16 @@
...
+from neural_compressor.metric import METRICS
 
 from .metrics import *
 
-METRICS = {
-    'pytorch': PytorchINCMetric,
-    'tensorflow': TensorflowINCMetric
+METRICS('pytorch').register('PytorchINCMetric', PytorchINCMetric)
+METRICS('tensorflow').register('TensorflowINCMetric', TensorflowINCMetric)
+
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/quantization/neural_compressor/metric/__init__.py:22]: Where `PytorchINCMetric` and `TensorflowINCMetric` are  defined?

File: python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py [hunk 2]
```diff
@@ -81,26 +83,9 @@ def post_training_quantize(self, model, calib_dataloader=None, val_dataloader=No
...
-                        to the new local class to avoid that.
-                        """
-                        self.metric = metric
-
-                self.metric = Metric(
-                    MyMetric,
-                    name="%s_%s" % (framework_metric.__name__, type(metric).__name__)
-                )
+                framework_inc_metric = INCMETRICS[self.cfg.model.framework]
+                framework_inc_metric.set_metric(metric)
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:87]: Does it make sense to modify a global state in a object? I think it is hard reason about the behavior and maintain the code. E.g. If I have a second `QuantizationINC` object with different metric, will it mess up the state of the first one?

Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:87]: You can't run them in parallel. Otherwise, this mess would be possible.

Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:87]: If you don't use multiple threads to do twice quantization in your assumption, that'll be fine. And seems doing that is not a good practice for post-training quantization.

File: python/nano/src/bigdl/nano/quantization/neural_compressor/metric/__init__.py (hunk 1 above)
Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/metric/__init__.py:22]: https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/nano/src/bigdl/nano/quantization/neural_compressor/metric/metrics.py#L19-L40

**Comment by zhentaocc** [comment]
Another option is to change the arguments into `quantize`.
Before:
```python
quantize(metric=torchmetrics.F1(10))
```
After:
```python
quantize(metric={torchmetrics.metrics.F1: {num_classes: 10}})
```
which is much more complicated for users to use. @yangw1234

File: python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py (hunk 2 above)
Comment (yangw1234) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:87]: I do not think you need multi-threading to mess it up. Does the following code work?
```python
quantA = QuantizationINC(metric-a)
quantB = QuantizationINC(metric-b)

quantA.some_method
quantB.some_method
```

Comment (yangw1234) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:87]: If there is a serialization problem, I think you should try to solve the serialization problem directly.

File: python/nano/src/bigdl/nano/quantization/neural_compressor/metric/base_metric.py [hunk 3]
```diff
@@ -24,13 +24,13 @@ def __init__(self):
...
+            return getattr(INCMetric, '_metric')
         except AttributeError:
             raise RuntimeError("Attribute 'metric' must be set.")
 
-    @metric.setter
-    def metric(self, framework_metric):
-        setattr(self, '_metric', framework_metric)
+    @staticmethod
+    def set_metric(framework_metric):
+        INCMetric._metric = framework_metric
```
[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/metric/base_metric.py:33]: Why can you dump a class variable but not a member variable? Is it because that a `TensorflowINCMetric` variable cannot be dumped?

File: python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py (hunk 2 above)
Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:87]: > I do not think you need multi-threading to mess it up. Does the following code work?
> 
> ```python
> quantA = QuantizationINC(metric-a)
> quantB = QuantizationINC(metric-b)
> 
> quantA.some_method
> quantB.some_method
> ```
Your use case is not actually correct. It should be
```python
trainer.quantize(model, trainloader, valloader, metric_a)
trainer.quantize(model, trainloader, valloader, metric_b)
```
or
```python
trainer.quantize(model, trainloader, valloader, metric_a)
trainer.quantize(model, trainloader, valloader, metric_b)
```python
quantA = QuantizationINC(**kwargs_a)
quantB = QuantizationINC(**kwargs_b)

quantized_1 = quantA.post_training_quantize(model, calib_dataloader, val_dataloader,
                                             metric_a)
quantized_2 = quantB.post_training_quantize(model, calib_dataloader, val_dataloader,
                                             metric_b)
quantized_3 = quantA.post_training_quantize(model, calib_dataloader, val_dataloader,
                                             metric_b)
quantized_4 = quantB.post_training_quantize(model, calib_dataloader, val_dataloader,
                                             metric_a)
```
With different kwargs and metric, the 4 quantization processes are all different and the metric wouldn't be messed up.

Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:87]: > If there is a serialization problem, I think you should try to solve the serialization problem directly.

If we want to resolve this directly, then metric pass to INC must be `{cls: kwargs}` and we need to assure that kwargs must be able to be serialized. So the solution is to pass a dict `{torchmetric_class: kwargs}` instead of instance of `torchmetric_class(kwargs)` as stated in comments below. Having discussed with @TheaperDeng , we agreed on passing a torchmetric instance which is more user-friendly.

Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:87]: > If there is a serialization problem, I think you should try to solve the serialization problem directly.

Or we need to check in some changes into INC directly... But this is not possible for the moment.

**Comment by jason-dai** [comment]
> resolves #4002 In previous implementation, each call of quantization, a metric will be registered. This causes error when we call quantization with a same metric in one script twice.
> 
> ### previous implementation :
> Each metric is registered at runtime, so there can be name conflict.
> 
> ```python
>   self.metric = Metric(
>       MyMetric,
>       name="%s_%s" % (framework_metric.__name__, type(metric).__name__)
>   )
> ```
> 
> ### After this fix:
> Customized metrics firstly are registered once for all statically.
> 
> ```python
> METRICS('pytorch').register('PytorchINCMetric', PytorchINCMetric)
> METRICS('tensorflow').register('TensorflowINCMetric', TensorflowINCMetric)
> ```
> 
> Then at runtime, we modify the real metric inside `INCMetric` by:
> 
> ```python
> PytorchINCMetric.set_metric(metric)
> ```
> 
> And fetch the registered metric by class name:
> 
> ```python
> self.cfg.evaluation.accuracy.metric = {PytorchINCMetric.__name__: None}
> ```
> 
> ### Modifications
> * [x]   Add default criterion to INC, allowing 0.99 acc drop by default
> * [x]   Fix metric registered in second time
> * [x]   Add test with metric assigned second time

I think this is too complex for solving a simple issue; to avoid conflict, you can simple append a unique id to the metrics name when registering it, something like:

```python
    self.metric = Metric(
       MyMetric,
       name="%s_%s_%d" % (framework_metric.__name__, type(metric).__name__, Metric.get_next_id())
   )
 ```

**Comment by zhentaocc** [comment]
> 
> I think this is too complex for solving a simple issue; to avoid conflict, you can simply append a unique id to the metrics name when registering it, something like:
> 
> ```python
>     self.metric = Metric(
>        MyMetric,
>        name="%s_%s_%d" % (framework_metric.__name__, type(metric).__name__, Metric.get_next_id())
>    )
> ```

Yes, this is much simpler.  
I was trying to refactor the code. I will firstly fix the issue, and open new PR to refactor if necessary.

**Comment by zhentaocc** [comment]
@jason-dai @yangw1234 @TheaperDeng How about current version?

File: python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py [hunk 4]
```diff
@@ -96,10 +96,9 @@ def __init__(self):
                         to the new local class to avoid that.
                         """
                         self.metric = metric
-
                 self.metric = Metric(
                     MyMetric,
-                    name="%s_%s" % (framework_metric.__name__, type(metric).__name__)
+                    name=f"nano_metric_{framework_metric.get_next_metric_id()}"
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:101]: give it a more meaningful name

Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:101]: yeah, maybe a metric name and an unique id.

Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/quantization.py:101]: OK, let me roll back the name.

File: python/nano/src/bigdl/nano/quantization/neural_compressor/metric/base_metric.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/metric/base_metric.py]: It's not about how the metric looks like, it's more about how the metric is created by INC. INC accepts a class `INCMetric` and a set of arguments `kwargs` to create a metric instance by INCMetric(**kwargs). And when we finished quantization in tensorflow, they try to dump the metric by a dict like `{INCMetric.name: kwargs}`. So it seems that we can't pass any argument that is not able to be dumped.  
### The problematic version
```python
self.metric=Metric(INCMetric, metric=metric)
```
Here, `kwargs={'metric': metric}` and `metric` can't be dumped.
### Resolve by class variable
```python
INCMetric.metric=metric
self.metric=Metric(INCMetric)
```
Here, `kwargs={}` , no issue.
### Resolve by local class inside function(current version)
```python
class MyMetric(INCMetric):
    def __init__(self):
        self.metric = metric
self.metric=Metric(MyMetric)
```
Here, `kwargs={}` , no issue.

[the remark this answers is not in the document]
Comment (zhentaocc) [reply python/nano/src/bigdl/nano/quantization/neural_compressor/metric/base_metric.py]: This issue only appears on tensorflow side. I think maybe pytorch and tensorflow have different postprocessings.

**Comment by jason-dai** [comment]
LGTM

**Comment by zhentaocc** [comment]
Jenkins passed http://10.112.231.51:18888/job/ZOO-PR-NanoTests/263/

**Review by yangw1234 — APPROVED** [review]

**Comment by jason-dai** [comment]
Also need to merge it to the main branch?

**Comment by TheaperDeng** [comment]
> Also need to merge it to the main branch?

I think main branch is created after this PR is merged. this commit is in main branch's tree.
